### ISSUE TITLE
Finding by owner username/group/uid/gid

### ISSUE THREAD

**Issue by alexmaco** [sharkdp/fd#307]
Hello,

I've been using `fd` for quite a long time, and loving it! Recently I've found myself needing to search for files owned by one user or another. I'm willing to do any implementation work required for it, and i'd like to start off a discussion and see if anyone else here has had a need for this feature.

Interface-wise, my first proposal would be to add a `--user` (as well as a `--group` switch) in the style of find, and maybe they could handle a uid/gid by default when passed a number.

Another idea, which is more flexible and less backwards-looking would be passing an `--owner` flag that can parse a `[(uid|username)][:(gid|groupname)]` input pair. As a precedent, docker's `--user` does this. Maybe it would make sense to call the flag `--user` in fd also, but i'm personally worried that it would make finding the "search by gid" option a little more obscure or unintuitive.

Lastly, I really don't know what to say right now about negative reasoning. I haven't encountered the need per se, but it feels "natural" in a way to be able to search for files "not owned by my user in this directory". Any ideas of how (or if) to do negation/exclusion by owner are welcome.

I'll start to hack a prototype together, and I'm really curious as what people think about this option.

**Comment by sharkdp** [comment]
Thank you very much for your feedback!

I think this request goes into a similar direction as #273 and #301 in the sense that I don't think that this is really part of the *"80% of the use cases"* (as mentioned in the README). That being said, I'm happy to consider adding this, if there is a strong support for this.

**Comment by genevera** [comment]
I just ran into this needing to find and delete directories owned by a certain UID inside of a huge dir tree

**Comment by sharkdp** [comment]
It looks like there *is* a strong support for this. Let's see if we can get #328 ready (some things need to be discussed first, see PR comments).

**Comment by alexmaco** [comment]
@sharkdp Good to hear that you are maybe willing to merge such a feature. I think now is a good point to discuss how it should look.

#328 currently up adds 4 cli flags, with no short forms. Since I filed the issue, I have been using fd from my fork (https://github.com/sharkdp/fd/compare/v7.5.0...alexmaco:find_by_owner) with the single `-o | --owner` option (which also supports negation) (that branch still has the `-U | --user` name, we can bikeshed here). I have come to find it quite usable, so i would suggest that the feature should be exposed like that:
* using a single option
* the argument should be `chown`-like (`(user|uid):(group|gid)`), with either side of the `:` being optional
* either side of `:` can be preceded with `!` to invert the match

That said, 4 options in ok too. Eventually `!` can be added to all 4 for negation. Also, someone needing more complicated file queries can use [fselect](https://github.com/jhspetersson/fselect) here instead. However I think the single option is cleaner, and it also mirrors syntax used by other CLI tools (like `chown` for uid|gid... , and `rip-grep` for the `!` negation).

The only downside I can see for the single argument version (aside from adding more syntax, which not everyone favors) is some possible user confusion. This may happen if the user has not encountered this kind of syntax before. For the help text, I have arrived at the following forms as a starting point, but feedback is welcome:
* for the short form: `-o, --owner <user:group>            Filter by user and/or group` 
* for the long form
```
    -o, --owner <user:group>               
            Filter files by their user and/or group.
            Format: (user|uid):(group|gid). Either side is optional.
            Precede either uid or gid with a ! to exclude files instead.
```

Since we have one PR up, i would not yet open another until maintainers decide on one approach or the other (and until i rebase my branch on v8.0.0 after all the refactoring :smile: )

**Comment by sharkdp** [comment]
@alexmaco Thank you very much for the detailed proposal!

> using a single option

Sounds great to me. Fewer is better :smile: 

> * the argument should be `chown`-like (`(user|uid):(group|gid)`), with either side of the `:` being optional
> 
> * either side of `:` can be preceded with `!` to invert the match

Sounds good to me as well. `!` is always a bit problematic as users have to remember to single-quote the argument. But I don't have a better idea. `~` has a similar problem (but works with double quotes).

> that branch still has the `-U | --user` name, we can bikeshed here

Quoting from #328:

> Although i initially though that --owner was a good idea in principle, especially since the argument can also take a group, i found that calling it --user still seems more natural, in the rule-of-least-astonishment way.

Why do you think `--user` would be more natural? Is there any precedent for that name? Otherwise, I would personally prefer `--owner`. But I agree. It's just bikeshedding.

> The only downside I can see for the single argument version (aside from adding more syntax, which not everyone favors) is some possible user confusion. This may happen if the user has not encountered this kind of syntax before.

Agreed. I don't think it's that bad. The `user:group` syntax should be familar to most users.

> For the help text, I have arrived at the following forms as a starting point, but feedback is welcome:
> […]
> Format: (user|uid):(group|gid). Either side is optional.

Hm. So the `:` has to be there? Or do we want to allow `--owner sharkdp` as well?

> Precede either uid or gid with a ! to exclude files instead.

uid => user/uid / gid => group/gid ?

**Comment by sharkdp** [comment]
> Since we have one PR up, i would not yet open another until maintainers decide on one approach or the other

I'd be in favor of your approach :+1: Just a small heads up: I might be unavailable for a review in the upcoming few weeks.

> and until i rebase my branch on v8.0.0 after all the refactoring smile

Sorry :see_no_evil:

**Comment by tmccombs** [comment]
> Hm. So the : has to be there? Or do we want to allow --owner sharkdp as well?

I think we should assume that if the : is missing it is a user/uid

**Comment by alexmaco** [comment]
> uid => user/uid / gid => group/gid ?

Yes

> Hm. So the `:` has to be there? Or do we want to allow `--owner sharkdp` as well?

Indeed, @tmccombs also spotted it: the `:` should not always be there.

I meant something akin to what `chown` accepts, so the complete grammar would be:
 `[ [!] (user|uid) ] [: [!] (group|gid) ]`

**Comment by sharkdp** [comment]
This has been implemented by @alexmaco in #581 :tada: 

The new functionality is currently not available on Windows.

**Comment by sharkdp** [comment]
Released in [fd 8.1.0](https://github.com/sharkdp/fd/releases/tag/v8.1.0):
``` bash
# files matching the 'exercise' pattern, owned by john
fd exercise --type file --owner john

# files/directories that are not owned by john
fd --owner '!john'

# … owned by the 'students' group
fd --owner ':students'

# … owned by the 'students' group, but not by john
fd --owner '!john:students'
```
